Calibration/HcalAlCaRecoProducers/src/AlCaIsoTracksProducer.cc

```diff
@@ -208,354 +208,249 @@ AlCaIsoTracksProducer::AlCaIsoTracksProducer(const edm::ParameterSet& iConfig)
   
-  ///vector of used hits:
-  std::vector<HcalDetId> usedHitsHC;
-  std::vector<int> usedHitsEC;
-  ///
-
-  // main loop over input tracks
-  for (reco::TrackCollection::const_iterator track=tC.begin(); track!=tC.end(); track++) {
-    bool noChargedTracks = true;
-    int itrk1=0;
-    itrk++;
-    double px = track->px();
-    double py = track->py();
-    double pz = track->pz();
-    double ptrack = sqrt(px*px+py*py+pz*pz);
+  //Step1: Get all the relevant containers
+  trigger::TriggerEvent triggerEvent;
+  edm::Handle<trigger::TriggerEvent> triggerEventHandle;
+  iEvent.getByToken(tok_trigEvt_, triggerEventHandle);
+  edm::Handle<edm::TriggerResults> triggerResults;
+  iEvent.getByToken(tok_trigRes_, triggerResults);
+
+  edm::Handle<reco::TrackCollection> trkCollection;
+  iEvent.getByToken(tok_genTrack_, trkCollection);
+  reco::TrackCollection::const_iterator trkItr;
+  edm::Handle<reco::VertexCollection> recVtxs;
+  iEvent.getByToken(tok_recVtx_, recVtxs);  
+  edm::Handle<reco::BeamSpot> beamSpotH;
+  iEvent.getByToken(tok_bs_, beamSpotH);
+  math::XYZPoint leadPV(0,0,0);
+  if (recVtxs->size()>0 && !((*recVtxs)[0].isFake())) {
+    leadPV = math::XYZPoint((*recVtxs)[0].x(),(*recVtxs)[0].y(),
+			    (*recVtxs)[0].z());
+  } else if (beamSpotH.isValid()) {
+    leadPV = beamSpotH->position();
+  }
+  LogDebug("HcalIsoTrack") << "Primary Vertex " << leadPV;
+
+  edm::Handle<EcalRecHitCollection> barrelRecHitsHandle;
+  edm::Handle<EcalRecHitCollection> endcapRecHitsHandle;
+  iEvent.getByToken(tok_EB_, barrelRecHitsHandle);
+  iEvent.getByToken(tok_EE_, endcapRecHitsHandle);
+  edm::Handle<HBHERecHitCollection> hbhe;
+  iEvent.getByToken(tok_hbhe_, hbhe);
+
+  //Get L1 trigger object
+  ptL1 = etaL1 = phiL1 = 0;
+  edm::Handle<trigger::TriggerFilterObjectWithRefs> l1trigobj;
+  iEvent.getByToken(tok_hltGT_, l1trigobj);
+  std::vector< edm::Ref<l1extra::L1JetParticleCollection> > l1tauobjref;
+  l1trigobj->getObjects(trigger::TriggerL1TauJet, l1tauobjref);
+  for (unsigned int p=0; p<l1tauobjref.size(); p++) {
+    if (l1tauobjref[p]->pt()>ptL1) {
+      ptL1  = l1tauobjref[p]->pt(); 
+      phiL1 = l1tauobjref[p]->phi();
+      etaL1 = l1tauobjref[p]->eta();
+    }
+  }
+  std::vector< edm::Ref<l1extra::L1JetParticleCollection> > l1jetobjref;
+  l1trigobj->getObjects(trigger::TriggerL1CenJet, l1jetobjref);
+  for (unsigned int p=0; p<l1jetobjref.size(); p++) {
+    if (l1jetobjref[p]->pt()>ptL1) {
+      ptL1  = l1jetobjref[p]->pt();
+      phiL1 = l1jetobjref[p]->phi();
+      etaL1 = l1jetobjref[p]->eta();
+    }
+  }
+  std::vector< edm::Ref<l1extra::L1JetParticleCollection> > l1forjetobjref;
+  l1trigobj->getObjects(trigger::TriggerL1ForJet, l1forjetobjref);
+  for (unsigned int p=0; p<l1forjetobjref.size(); p++) {
+    if (l1forjetobjref[p]->pt()>ptL1) {
+      ptL1  = l1forjetobjref[p]->pt();
+      phiL1 = l1forjetobjref[p]->phi();
+      etaL1 = l1forjetobjref[p]->eta();
+    }
+  }
+  //For valid HLT record
```

Comment:
These three "for" loops for ptL1, phiL1, and etaL1 are almost identical. The code would be easier to understand if the loop was turned into a function that was called three times, rather than having a block of code repeated three times.